include/arch/arm/aarch64/asm_inline_gcc.h

```diff
@@ -48,21 +48,19 @@ static ALWAYS_INLINE unsigned int arch_irq_lock(void)
 	 */
 	__asm__ volatile("mrs %0, daif;"
 			 "msr daifset, %1;"
-			 "isb"
 			 : "=r" (key)
 			 : "i" (DAIFSET_IRQ)
-			 : "memory", "cc");
+			 : "memory");
```

Comment:
Writes to it may do occur in program order, but this does not prevent subsequent instructions to be fetched and/or executed before the write is executed, that's why I guess you had added the "ISB" here in the first place. I am not so familiar with the intrinsics of the the AARCH64, so I just bring this up.

Reply:
Anything that relied on memory ordering across `arch_irq_lock` was already likely broken, unfortunately. `isb` does not order memory accesses.

As for instructions themselves, being executed out of order doesn't matter here, given the architectural guarantee that daif writes execute in program order.

There are many other system registers that don't guarantee ordering without an `isb` (or other ordering event, in some cases). `daifset` is one of the few exceptions there.

Reply:
Architecture guarantees that the writes execute in program order, i.e. are observed before other writes are completed; without ISB we wouldn't normally guarantee the writes are realized before any subsequent instruction is fetched/decoded/executed. Maybe something is special for PSTATE, I am reading this in the spec:

`Writes to PSTATE.{PAN, D, A, I, F} occur in program order without the need for additional synchronization. `

Does this imply also need for ISB synchronization? If so, then the patch is fine.
@carlocaione pls, confirm this.

Reply:
This is exactly what I reported in the commit message and that explain why the ISB is not needed in this case, so the patch is fine according to my interpretation of the ARM manual.

Reply:
OK, spec is not 100% clear to me about this, but fine if you confirm. 
FWIW, in ARM M profile, only CPSID is guaranteed to work without ISB (CPSIE needs it).

Reply:
Well, this is confusing but looking at the A-TF sources there the `isb()` is used indeed. @jharris-intel what's your take on this?

Reply:
So, the ISB here _does_ do something, but it's not normally necessary, and is a bit of a pessimization unless you need that particular behavior.

In particular, under ARM (AArch32 and AArchh64), interrupts are guaranteed to fire within a finite time (assuming unmasked), but are otherwise not guaranteed to fire until the next synchronization event (any of a number of things, notably including ISB).

So, what this means is roughly as follows. If you were previously relying on an `irq unlock` to _guarantee_ that a pending interrupt fires _immediately_, as opposed to firing "at it's leisure", then yes, you need the ISB. However, this is at the expense of a nontrivial stall (ISBs are not the cheapest instructions in the world) in a very common operation.

...although 99% of code that requires that an interrupt fires immediately in this situation is broken _anyway_. May be slightly hyperbolic, but  not by much.

Reply:
> So, what this means is roughly as follows. If you were previously relying on an irq unlock to guarantee that a pending interrupt fires immediately, as opposed to firing "at it's leisure", then yes, you need the ISB. However, this is at the expense of a nontrivial stall (ISBs are not the cheapest instructions in the world) in a very common operation.

This is the point I am trying to make here, determinism. :)
I think the ISB should be brought back, to make sure IRQs can _always_ be fired immediately after the unlock(); instruction, assuming they are properly unmasked, as @jharris-intel mentions.

Reply:
@ioannisg - unfortunately, if that's your requirement then you're already out of luck :-(

The GIC itself has no guarantees about time to process an interrupt (and indeed in some cases especially with LPIs this can be a nontrivial amount of time.)

Reply:
@jharris-intel, I am actually talking about a very specific scenario, which is:
If an IRQ is already pending, while the interrupts are locked, I would like this IRQ to fire **before** any instruction gets fetched, after the user has called irq_unlock(key). [Assuming of course that the unlock() un-masks the pending interrupt].

Cortex-M use case is as follows:

In Cortex-M the CPSIE guarantees the above. In mainline Cortex-M, however, we use BASEPRI to unlock interrupts, and the architecture spec does not guarantee that synchronization is implicit to the BASEPRI setting (via MSR), so we do an ISB to be 100% sure.

I am wondering if Cortex-A needs something similar.

Reply:
> If an IRQ is already pending, while the interrupts are locked, I would like this IRQ to fire **before** any instruction gets fetched, after the user has called irq_unlock(key).

Is this an hard requirement for Zephyr / `irq_unlock(key)` or just something nice-to-have? Or there is any other reason you want to have the IRQ to fire before any instruction is fetched?

Reply:
I can't tell if this is a hard requirement, by looking at the irq_unlock documentation. I guess it is reasonable to aim at determinism, if we can have that. Then, also the argument of uniformity/consistency across arches. But nothing more than that.

Reply:
I think we no need ISB here. Checked Linux Kernel code, no ISB.

Reply:
> I am wondering if Cortex-A needs something similar.

Offhand, no. Or rather:

You bring up a good point, in that determinism here is arguably desirable.

On Cortex-As this comes at a significant cost (ISBs aren't exactly cheap), and doesn't achieve the objective (the GIC has no guarantees about interrupt latency).

*****

It may be worth at some point introducing a `CONFIG_PREFER_DETERMINISM_OVER_PERFORMANCE` flag (or rather, a better-named flag). There are a surprising number of knobs that can be tweaked here (`ISB` when disabling interrupts, restartable LDM/STM on AArch32, etc, etc).